File: src/brs/web/api/http/handler/GetAccountId.java
package brs.web.api.http.handler;

import static brs.web.api.http.common.JSONResponses.MISSING_SECRET_PHRASE_OR_PUBLIC_KEY;
import static brs.web.api.http.common.Parameters.PUBLIC_KEY_PARAMETER;
import static brs.web.api.http.common.Parameters.SECRET_PHRASE_PARAMETER;
import static brs.web.api.http.common.ResultFields.ACCOUNT_RESPONSE;
import static brs.web.api.http.common.ResultFields.PUBLIC_KEY_RESPONSE;

import javax.servlet.http.HttpServletRequest;

import brs.web.api.http.ApiServlet;
import brs.web.api.http.common.JSONData;
import brs.web.api.http.common.LegacyDocTag;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import brs.crypto.Crypto;
import brs.util.Convert;
import signumj.crypto.SignumCrypto;
import signumj.entity.SignumAddress;

public final class GetAccountId extends ApiServlet.JsonRequestHandler {

  static final GetAccountId instance = new GetAccountId();

  public GetAccountId() {
    super(new LegacyDocTag[] {LegacyDocTag.ACCOUNTS}, SECRET_PHRASE_PARAMETER, PUBLIC_KEY_PARAMETER);
  }

  @Override
  protected
  JsonElement processRequest(HttpServletRequest req) {

    SignumAddress address;
    String secretPhrase = Convert.emptyToNull(req.getParameter(SECRET_PHRASE_PARAMETER));
    String publicKeyString = Convert.emptyToNull(req.getParameter(PUBLIC_KEY_PARAMETER));
    if (secretPhrase != null) {
      byte[] publicKey = Crypto.getPublicKey(secretPhrase);
      address = SignumCrypto.getInstance().getAddressFromPublic(publicKey);
      publicKeyString = Convert.toHexString(publicKey);
    } else if (publicKeyString != null) {
      address = SignumCrypto.getInstance().getAddressFromPublic(Convert.parseHexString(publicKeyString));
    } else {
      return MISSING_SECRET_PHRASE_OR_PUBLIC_KEY;
    }

    JsonObject response = new JsonObject();
    JSONData.putAccount(response, ACCOUNT_RESPONSE, address.getSignedLongId());
    response.addProperty(ACCOUNT_RESPONSE + "RSExtended", address.getExtendedAddress());
    response.addProperty(PUBLIC_KEY_RESPONSE, publicKeyString);

    return response;
  }
  boolean requirePost() {
    return true;
  }

}

File: src/main/kotlin/brs/api/http/GetAccountId.kt
package brs.api.http

import brs.api.http.common.JSONResponses.MISSING_SECRET_PHRASE_OR_PUBLIC_KEY
import brs.api.http.common.Parameters.PUBLIC_KEY_PARAMETER
import brs.api.http.common.Parameters.SECRET_PHRASE_PARAMETER
import brs.api.http.common.ResultFields.ACCOUNT_RESPONSE
import brs.api.http.common.ResultFields.PUBLIC_KEY_RESPONSE
import brs.api.http.common.JSONData
import brs.util.convert.emptyToNull
import brs.util.convert.parseHexString
import brs.util.convert.publicKeyToId
import brs.util.convert.toHexString
import brs.util.crypto.Crypto
import brs.util.jetty.get
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import javax.servlet.http.HttpServletRequest

/**
 * TODO
 */
internal class GetAccountId :
    APIServlet.JsonRequestHandler(arrayOf(APITag.ACCOUNTS), SECRET_PHRASE_PARAMETER, PUBLIC_KEY_PARAMETER) {
    override fun processRequest(request: HttpServletRequest): JsonElement {

        val accountId: Long
        val secretPhrase = request[SECRET_PHRASE_PARAMETER].emptyToNull()
        var publicKeyString = request[PUBLIC_KEY_PARAMETER].emptyToNull()
        when {
            secretPhrase != null -> {
                val publicKey = Crypto.getPublicKey(secretPhrase)
                accountId = publicKey.publicKeyToId()
                publicKeyString = publicKey.toHexString()
            }
            publicKeyString != null -> accountId = publicKeyString.parseHexString().publicKeyToId()
            else -> return MISSING_SECRET_PHRASE_OR_PUBLIC_KEY
        }

        val response = JsonObject()
        JSONData.putAccount(response, ACCOUNT_RESPONSE, accountId)
        response.addProperty(PUBLIC_KEY_RESPONSE, publicKeyString)

        return response
    }

    override fun requirePost(): Boolean {
        return true
    }

    companion object {

        internal val instance = GetAccountId()
    }

}
